### PR TITLE
Rag pinecone

### PR DESCRIPTION
**Pull request by brianpark0126**

### CODE REVIEW

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+chat = OpenAIChat()
+embeddings_model = "text-embedding-ada-002"
+
+if not os.path.exists("cookbook/basic_rag_example/news_article_dataset.csv"):
+    url = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
+    df = pd.read_csv(url)
+    split_articles = []
+    encoder = tiktoken.encoding_for_model("gpt-3.5-turbo")
+
+    for i, row in df.iterrows():
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:26]: make this into a helper function with a docstring that explains it

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import tiktoken
+from pinecone import Pinecone, ServerlessSpec
+
+from mirascope import OpenAIChat, Prompt
+
+os.environ["OPENAI_API_KEY"] = "YOUR_API_KEY"
+chat = OpenAIChat()
+embeddings_model = "text-embedding-ada-002"
+
+if not os.path.exists("cookbook/basic_rag_example/news_article_dataset.csv"):
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:20]: This loading should read like this:

```python
from urllib.request import urlretrieve

url = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
filename = "news_article_dataset.csv"  # reading from pwd basic_rag_example/
if not os.path.exists(filename):
    urlretrieve(url, filename)

df = pd.read_csv(filename)
...
```

File: cookbook/basic_rag_example/pinecone_embeddings.py (hunk 1 above)
Comment (willbakst) [reply cookbook/basic_rag_example/pinecone_embeddings.py:26]: this should be something like `df = new_function(df)`

Covers lines 26-40

Also make sure to add a comment explaining lines 40-42 in the new function

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            ]
+            split_articles += split_texts
+            df.drop(i, inplace=True)
+
+    df = pd.concat(
+        [df, pd.DataFrame(split_articles, columns=["Text"])], ignore_index=True
+    )
+    embeddings = []
+    # ada 02 supports ~8000 tokens and 8 articles fit in the max batch size
+    for start_idx in range(0, len(df), 8):
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:45]: either `start_index` or `start`

Comment (willbakst) [reply cookbook/basic_rag_example/pinecone_embeddings.py:45]: generally do not abbreviate variables

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        metric="cosine",
+        spec=ServerlessSpec(
+            cloud="aws",
+            region="us-west-2",
+        ),
+    )
+index = pc.Index(index_name)
+
+
+batch_size = len(df) // 2
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:76]: what's going on here with the divide by 2?

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        .data[0]
+        .embedding
+    )
+    articles = []
+    for batch in ["batch1", "batch2"]:
+        query_response = index.query(namespace=batch, vector=query_embedding, top_k=2)
+        relevant_article_indices = [
+            int(article["id"]) for article in query_response["matches"]
+        ]
+        articles += relevant_article_indices
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:116]: this should be:

```python
article_indices += [int(article["id"]) for article in query_response["matches"]]
```

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    relevant_articles = df.loc[articles]["Text"].tolist()
+    prompt = NewsPrompt(topic=query, articles=relevant_articles)
+    res = chat.create(prompt)
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:119]: change `res` -> `completion`

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+def summarize_news(query):
+    """Summarizes 2004 news about retrieved context relevant to query."""
+    query_embedding = (
+        chat.client.embeddings.create(model=embeddings_model, input=[query])
+        .data[0]
+        .embedding
+    )
+    articles = []
+    for batch in ["batch1", "batch2"]:
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:111]: this as above is not clear, so please explain somewhere why we have two batches, why not 1? why not 4?

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+embeddings_model = "text-embedding-ada-002"
+
+
+def load_data(url) -> pd.DataFrame:
+    """Loads data from a url, and splits larger texts into smaller chunks."""
+    df = pd.read_csv(url)
+    split_articles = []
+    encoder = tiktoken.encoding_for_model("gpt-3.5-turbo")
+
+    def split_text(text, tokens, max_tokens):
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:27]: move this outside of load_data as it's own function, add type hints to the parameters, and flesh out the docstring to be a full docstring

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        [df, pd.DataFrame(split_articles, columns=["Text"])], ignore_index=True
+    )
+
+    return df
+
+
+def embed_data(df, batch_size):
+    """Embeds a dataframe of texts and stores it in a new column."""
+    embeddings = []
+    # ada 02 supports ~8000 tokens and 8 articles fit in the max batch size
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:60]: this comment is duplicated. I would suggest removing it here and leaving it below

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        )
+        embeddings += [datum.embedding for datum in embeddings_response.data]
+    df["embeddings"] = embeddings
+
+    return df
+
+
+filename = "news_article_dataset.csv"
+# Make reruns efficient by preprocessing data before saving to csv
+if not os.path.exists("cookbook/basic_rag_example/news_article_dataset.csv"):
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:75]: this should just be `filename` since you should be running this from the `basic_rag_example/` directory (not base mirascope).

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    url = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
+    df = load_data(url)
+    df = embed_data(df, 8)  # ada 02 supports ~8000 tokens, our max_tokens is 1000
+    df.to_csv(f"{os.getcwd()}/cookbook/basic_rag_example/{filename}")
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:79]: same this here. You should just be able to write `df.to_csv(filename)`

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+else:
+    df = pd.read_csv(f"{os.getcwd()}/cookbook/basic_rag_example/{filename}")
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:81]: same here

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        # Embeddings need to be parsed if they've been read from a csv
+        if isinstance(embeddings[0], str):
+            embeddings = ast.literal_eval(embeddings)
+        vectors.append({"id": str(i), "values": embeddings})
+    index.upsert(vectors=vectors, namespace="articles")
+
+
+class NewsPrompt(Prompt):
+    """
+    Here are 4 relevant sections of news articles relevant to this topic: {topic}.
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:116]: update 4 here to be `{num_articles}` and then add a property that is the length of the articles list

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        split_texts = [
+            " ".join(words[i : i + words_per_split])
+            for i in range(0, len(words), words_per_split + 1)
+        ]
+        return split_texts
+
+    for i, row in df.iterrows():
+        text = row["Text"]
+        tokens = encoder.encode(text)
+        if len(tokens) > 1000:
```
Comment (brenkao) [inline cookbook/basic_rag_example/pinecone_embeddings.py:45]: define variable max_tokens and set to 1000, and use it on this line and the other below

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+    # Long texts which were dropped from the dataframe are now readded.
+    df = pd.concat(
+        [df, pd.DataFrame(split_articles, columns=["Text"])], ignore_index=True
+    )
+
+    return df
+
+
+def embed_data(df, batch_size):
```
Comment (brenkao) [inline cookbook/basic_rag_example/pinecone_embeddings.py:57]: add return type

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        num_splits = len(tokens) // max_tokens + 1
+        words_per_split = len(words) // num_splits
+        split_texts = [
+            " ".join(words[i : i + words_per_split])
+            for i in range(0, len(words), words_per_split + 1)
+        ]
+        return split_texts
+
+    for i, row in df.iterrows():
+        text = row["Text"]
```
Comment (brenkao) [inline cookbook/basic_rag_example/pinecone_embeddings.py:43]: use variable for "Text" and replace all instances of "Text"

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        .embedding
+    )
+    query_response = index.query(namespace="articles", vector=query_embedding, top_k=4)
+    relevant_article_indices = [
+        int(article["id"]) for article in query_response["matches"]
+    ]
+    relevant_articles = df.loc[relevant_article_indices]["Text"].tolist()
+    prompt = NewsPrompt(topic=query, articles=relevant_articles)
+    completion = chat.create(prompt)
+    print(str(prompt))
```
Comment (brenkao) [inline cookbook/basic_rag_example/pinecone_embeddings.py]: print(prompt), lets aim to remove all unnecessary castings

File: pyproject.toml [hunk 18]
```diff
@@ -38,6 +38,8 @@ fastapi = "^0.105.0"
 numpy = "^1.26.3"
 pandas = "^2.2.0"
 pandas-stubs = "^2.1.4.231227"
+tiktoken = "^0.5.2"
```
Comment (brenkao) [inline pyproject.toml:41]: why are these added to our pyproject, instead of the requirements.txt

File: pyproject.toml [hunk 19]
```diff
@@ -38,6 +38,8 @@ fastapi = "^0.105.0"
...
+pinecone-client = "^3.0.1"
```
Comment (brenkao) [inline pyproject.toml:42]: ditto

File: cookbook/basic_rag_example/pinecone_embeddings.py (hunk 16 above)
Comment (willbakst) [reply cookbook/basic_rag_example/pinecone_embeddings.py:52]: use `TEXT_COLUMN = "Text"` somewhere at the top and then use `TEXT_COLUMN` everywhere

File: cookbook/basic_rag_example/pinecone_embeddings.py (hunk 14 above)
Comment (willbakst) [reply cookbook/basic_rag_example/pinecone_embeddings.py:54]: same here i.e. `MAX_TOKENS = 1000`

File: pyproject.toml (hunk 18 above)
Comment (willbakst) [reply pyproject.toml:41]: it's added to both. weirdly github smushed this, if you expand you'll see this is added to the cookbook dependency group (I almost wrote the same comment)

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    ]
+
+    return split_texts
+
+
+def load_data(url: str) -> pd.DataFrame:
+    """Loads data from a url, and splits larger texts into smaller chunks."""
+    df = pd.read_csv(url)
+    split_articles = []
+    encoder = tiktoken.encoding_for_model("gpt-3.5-turbo")
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:52]: similarly extract this into something like `MODEL = "gpt-3.5-turbo"` and then use it above for `chat = OpenAIModel(model=MODEL)`

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+        embeddings_response = chat.client.embeddings.create(
+            model=embeddings_model, input=batch
+        )
+        embeddings += [datum.embedding for datum in embeddings_response.data]
+    df[EMBEDDINGS_COLUMN] = embeddings
+
+    return df
+
+
+filename = "news_article_dataset.csv"
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:84]: maybe also pull these above into `FILENAME` and `URL`

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+import pandas as pd
+import tiktoken
+from pinecone import Pinecone, ServerlessSpec
+
+from mirascope import OpenAIChat, Prompt
+
+os.environ["OPENAI_API_KEY"] = "YOUR_API_KEY"
+chat = OpenAIChat()
+embeddings_model = "text-embedding-ada-002"
+TEXT_COLUMN = "Text"
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:19]: group these better:

```python
os.environ["OPENAI_API_KEY"] = "YOUR_API_KEY"

MODEL = "gpt-3.5-turbo"
TEXT_COLUMN = "Text"
EMBEDDINGS_COLUMN = "embeddings"
EMBEDDINGS_MODEL = "text-embedding-ada-002"

chat = OpenAIChat(model=MODEL)
```

**Review by willbakst — APPROVED** [review]

File: cookbook/basic_rag_example/pinecone_embeddings.py [hunk 23]
```diff
@@ -14,17 +14,25 @@
...
-embeddings_model = "text-embedding-ada-002"
+
+MODEL = "gpt-3.5-turbo"
+EMBEDDINGS_MODEL = "text-embedding-ada-002"
 TEXT_COLUMN = "Text"
 EMBEDDINGS_COLUMN = "embeddings"
 
+chat = OpenAIChat(model=MODEL)
+
+filename = "news_article_dataset.csv"
```
Comment (willbakst) [inline cookbook/basic_rag_example/pinecone_embeddings.py:25]: also make these all caps constants `FILENAME` and `URL`

**Review by willbakst — APPROVED** [review]
